### DISCUSSION TITLE
Compilation on cluster

### DISCUSSION THREAD

**Discussion by selarem** [idaholab/moose#25439] [Q&A General]
Hello,
I am trying to compile moose on a cluster. I scp the version I have on my computer which compiles normally under ubuntu.
When trying to compile it on the cluster I obtain:
**********
File "/lus/home/CT9/cin4054/elarem/projects/moose/framework/scripts/get_repo_revision.py", line 19, in <module>
    from packaging import version
ModuleNotFoundError: No module named 'packaging'
Checking if header needs updating: /lus/home/CT9/cin4054/elarem/projects/moose/modules/combined/include/base/CombinedRevis
ion.h...
      Successfully remade target file '/lus/home/CT9/cin4054/elarem/projects/moose/modules/combined/include/base/CombinedR
evision.h'.
         File 'combined_header_symlinks' does not exist.
        Must remake target 'combined_header_symlinks'.
        Successfully remade target file 'combined_header_symlinks'.
     Prerequisite '/lus/home/CT9/cin4054/elarem/projects/moose/modules/combined/include/base/CombinedRevision.h' of target
 '/lus/home/CT9/cin4054/elarem/projects/moose/modules/combined/lib/libcombined-opt.la' does not exist.
    Must remake target '/lus/home/CT9/cin4054/elarem/projects/moose/modules/combined/lib/libcombined-opt.la'.
Linking Library /lus/home/CT9/cin4054/elarem/projects/moose/modules/combined/lib/libcombined-opt.la...
    Successfully remade target file '/lus/home/CT9/cin4054/elarem/projects/moose/modules/combined/lib/libcombined-opt.la'.
     File '/lus/home/CT9/cin4054/elarem/projects/moose/modules/combined/test/lib/libcombined_test-opt.la' does not exist.
       File '/lus/home/CT9/cin4054/elarem/projects/moose/modules/combined/test/src/base/CombinedTestApp.x86_64-pc-linux-gn
u.opt.lo' does not exist.
         File '/lus/home/CT9/cin4054/elarem/projects/moose/modules/combined/build/header_symlinks/CombinedRevision.h' does
 not exist.
        Must remake target '/lus/home/CT9/cin4054/elarem/projects/moose/modules/combined/build/header_symlinks/CombinedRev
ision.h'.
        Successfully remade target file '/lus/home/CT9/cin4054/elarem/projects/moose/modules/combined/build/header_symlink
s/CombinedRevision.h'.
      Must remake target '/lus/home/CT9/cin4054/elarem/projects/moose/modules/combined/test/src/base/CombinedTestApp.x86_6
4-pc-linux-gnu.opt.lo'.
Compiling C++ (in opt mode) /lus/home/CT9/cin4054/elarem/projects/moose/modules/combined/test/src/base/CombinedTestApp.C..
.
/lus/home/CT9/cin4054/elarem/projects/moose/modules/combined/test/src/base/CombinedTestApp.C:16:10: fatal error: CombinedR
evision.h: No such file or directory
 #include "CombinedRevision.h"
          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
*****
the script used is :
#!/bin/bash
#SBATCH --account=cin4054
#SBATCH --time=00:10:00
#SBATCH  -J cyl-Re-3900-vis431-6h
#SBATCH --constraint=GENOA
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --ntasks-per-core=1 # 
#SBATCH --threads-per-core=1 # --hint=nomultithread
#SBATCH --exclusive
#SBATCH --output=labelled-test-%J.out
#SBATCH --error=labelled-test-%J.err

module purge
source /opt/cray/pe/cpe/23.02/restore_lmod_system_defaults.sh
module load PrgEnv-gnu
module load craype-x86-genoa
module load cray-python
module load craype-network-ofi
module load CPE-23.02-gcc-12.2.0-CPU-softs
module list

make

Thank you for any help.
regards,
Sab

**Comment by GiudGiud** [comment]
Hello

you’re missing the packaging Python package. You can download it with mamba for example

Guillaume

**Reply by alimostafavi24** [reply]
Hello,

I also have the same problem. I used module load python/3.8.12. How can I add the packaging Python package on the sdsc HPC system?

Thanks,
Ali

**Reply by GiudGiud** [reply]
Hello

you can download it through conda or through pip
You could ask your HPC team to add it to the local Python install 

Guillaume

**Reply by selarem** [reply]
On cluster it is not that evident without internet connexion. Should go with Offline Installation. I am done with compiling moose, but for my APP, which is running fine on my station, I obtain (on cluster):

```
lus/home/CT9/cin4054/elarem/projects/allpf/build/header_symlinks/StressSpectralQDC.h:55:34: error: 'StressSpectralQDC::_gc' will be initialized after [-Werror=reorder]
   55 |   const MaterialProperty<Real> & _gc;
      |                                  ^~~
/lus/home/CT9/cin4054/elarem/projects/allpf/build/header_symlinks/StressSpectralQDC.h:50:23: error:   'const VariableValue& StressSpectralQDC::_c' [-Werror=reorder]
   50 | const VariableValue & _c;
      |                       ^~
/lus/home/CT9/cin4054/elarem/projects/allpf/src/materials/StressSpectralQDC.C:51:1: error:   when initialized here [-Werror=reorder]
   51 | StressSpectralQDC::StressSpectralQDC(const InputParameters & parameters)
      | ^~~~~~~~~~~~~~~~~
In file included from /lus/home/CT9/cin4054/elarem/projects/allpf/src/materials/StressSpectralQDNC.C:10,
                 from /lus/home/CT9/cin4054/elarem/projects/allpf/build/unity_src/materials_Unity.C:20:
/lus/home/CT9/cin4054/elarem/projects/allpf/build/header_symlinks/StressSpectralQDNC.h: In constructor 'StressSpectralQDNC::StressSpectralQDNC(const InputParameters&)':
/lus/home/CT9/cin4054/elarem/projects/allpf/build/header_symlinks/StressSpectralQDNC.h:70:31: error: 'StressSpectralQDNC::_deltaMixed' will be initialized after [-Werror=reorder]
   70 |   RankFourTensor _deltaOuter, _deltaMixed;
      |                               ^~~~~~~~~~~
/lus/home/CT9/cin4054/elarem/projects/allpf/build/header_symlinks/StressSpectralQDNC.h:51:6: error:   'libMesh::Real StressSpectralQDNC::_kdamage' [-Werror=reorder]
   51 | Real _kdamage;
```

It is running without problems on my computer.

regards,

**Reply by GiudGiud** [reply]
hello

you just need to re-order the attributes in the constructor to remove this error

you're getting that because we pass the "all errors" compilation flag and this is a warning

Guillaume

**Reply by selarem** [reply]
So, I need to change the order in the constructor to correspond to what in the header file?  Is there a way to shut these warning that are, on this cluster, considered as errors ? On my station the same App compiles fine. 
Saber

**Reply by GiudGiud** [reply]
You can modify the CXX FLAGS to allow warnings, and especially that warning, but it s easier to fix it once and for all

**Comment by ghost** [comment]
> You can modify the CXX FLAGS to allow warnings, and especially that warning, but it s easier to fix it once and for all

Hello, 

So we've fixed all of the reorder errors and this new error shows up now for all files : 
```
`In file included from /lus/home/CT9/cin4054/elarem/projects/allpf/build/unity_src/materials_Unity.C:19:
/lus/home/CT9/cin4054/elarem/projects/allpf/src/materials/SqrtQDC.C:16:1: error: redefinition of 'static InputParameters SqrtQDC::validParam
   16 | SqrtQDC::validParams()
      | ^~~~~~~
In file included from /lus/home/CT9/cin4054/elarem/projects/allpf/build/unity_src/materials_Unity.C:18:
/lus/home/CT9/cin4054/elarem/projects/allpf/src/materials/SqrtQDC.C:16:1: note: 'static InputParameters SqrtQDC::validParams()' previously d
   16 | SqrtQDC::validParams()
      | ^~~~~~~
/lus/home/CT9/cin4054/elarem/projects/allpf/src/materials/SqrtQDC.C:41:1: error: redefinition of 'SqrtQDC::SqrtQDC(const InputParameters&)'
   41 | SqrtQDC::SqrtQDC(const InputParameters & parameters)
      | ^~~~~~~
/lus/home/CT9/cin4054/elarem/projects/allpf/src/materials/SqrtQDC.C:41:1: note: 'SqrtQDC::SqrtQDC(const InputParameters&)' previously define
   41 | SqrtQDC::SqrtQDC(const InputParameters & parameters)
      | ^~~~~~~
/lus/home/CT9/cin4054/elarem/projects/allpf/src/materials/SqrtQDC.C:75:1: error: redefinition of 'void SqrtQDC::initQpStatefulProperties()'
   75 | SqrtQDC::initQpStatefulProperties()
      | ^~~~~~~
/lus/home/CT9/cin4054/elarem/projects/allpf/src/materials/SqrtQDC.C:75:1: note: 'virtual void SqrtQDC::initQpStatefulProperties()' previousl
   75 | SqrtQDC::initQpStatefulProperties()
      | ^~~~~~~
/lus/home/CT9/cin4054/elarem/projects/allpf/src/materials/SqrtQDC.C:83:1: error: redefinition of 'void SqrtQDC::computeQpStress()'
   83 | SqrtQDC::computeQpStress()
      | ^~~~~~~
/lus/home/CT9/cin4054/elarem/projects/allpf/src/materials/SqrtQDC.C:83:1: note: 'virtual void SqrtQDC::computeQpStress()' previously defined
   83 | SqrtQDC::computeQpStress()
      | ^~~~~~~
/lus/home/CT9/cin4054/elarem/projects/allpf/src/materials/SqrtQDC.C:231:1: error: redefinition of 'void SqrtQDC::returnMap(const RankTwoTensconst RankFourTensor&, RankTwoTensor&, libMesh::Real&, RankTwoTensor&)'
  231 | SqrtQDC::returnMap(const RankTwoTensor & sig_old,
      | ^~~~~~~
/lus/home/CT9/cin4054/elarem/projects/allpf/src/materials/SqrtQDC.C:231:1: note: 'virtual void SqrtQDC::returnMap(const RankTwoTensor&, libMkFourTensor&, RankTwoTensor&, libMesh::Real&, RankTwoTensor&)' previously defined here
  231 | SqrtQDC::returnMap(const RankTwoTensor & sig_old,
      | ^~~~~~~

`
```

**Reply by GiudGiud** [reply]
hello

this happens when you have two source files that define the same class.
Happens to me all the time when I copy a source file then forget to adapt it
